FIX adjust inliner criterion in RANSACRegressor

gregorystrubel opened the pull request:
#### Reference Issues/PRs
Fixes #19497. 
Fixes #16147.

#### What does this implement/fix? Explain your changes.
This fix implement solution discussed in #16147: 
inlier_mask_subset = residuals_subset <= residual_threshold
that replace the current <

This make the code work also with perfect horizontal lines (see #19497). 

Basically I suggest to change the definition of what is an inlier respectivly to residual_threshold. Exactly on the threshold, I suggest to say it is an inlier and not an outliner.

This make a couple of "normal cases" work: horizontal line #19497 (currently we say all the point on a line are outlier!) and other examples described in #16147

Unfortunately This patch make a test fail, but we can keep it with the following modification. The test construct an example where there is no inlier. It is based on the fact when residual_threshold =0 an exception is always thrown. To do this in a clean way we should construct here a real example where there is no inlier. Such an example is not easy to find. 
I suggest to construct a more simple test as before : set the residual_threshold to an other corner value that make always the comparison fail : a nan

I also implement the test described in #19497 (horizontal lines)
#### Any other comments?

Commit:
bugfix RANSAC - residual threshold calculation #16147

Commit:
RANSAC test

add new test for horizontal line
adapt no inliner test with a synonym

Commit:
flake8 correction

ogrisel commented:
Thanks for the fix, the detailed explanation and the new test.

Can you please check if the definition of an inlier needs to be updated in the documentation? If not it might be worth making the doc a bit more specific (either in the user guide or the docstring or both).

Can you also please document your fix in the changelog (`doc/what_new/v1.0.rst`)?

ogrisel commented on sklearn/linear_model/tests/test_ransac.py:
```diff
@@ -531,3 +531,15 @@ def test_ransac_final_model_fit_sample_weight():
     assert_allclose(ransac.estimator_.coef_, final_model.coef_, atol=1e-12)
+
+
+def test_perfect_horizontal_line():
```
Please add a comment to explain that this is a not regression test for `https://github.com/scikit-learn/scikit-learn/issues/16147`.

gregorystrubel replied:
Thanks a lot for your review. I add a comment to explain this is a regression test for #19497.
I consider your "not" as a typo. If it is not, could you explain me why it is not a regression test ?

jjerphan replied:
I think @ogrisel meant "a non-regression test", hence a typo.

gregorystrubel replied:
thanks

Commit:
documentation changes

Commit:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into bugfix_ransac_residual_threshold

gregorystrubel commented:
Thanks a lot for your review. I have updated the documentation.

Commit:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into bugfix_ransac_residual_threshold

Commit:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into bugfix_ransac_residual_threshold

jjerphan commented on doc/modules/linear_model.rst:
```diff
@@ -1265,8 +1265,8 @@ Each iteration performs the following steps:
    to the estimated model (``base_estimator.predict(X) - y``) - all data
-   samples with absolute residuals smaller than the ``residual_threshold``
-   are considered as inliers.
+   samples with absolute residuals smaller or equal than the 
+   ``residual_threshold`` are considered as inliers.
```
Small syntax fix due to the addition.
```suggestion
   samples with absolute residuals smaller than or equal to the 
   ``residual_threshold`` are considered as inliers.
```

jjerphan commented on doc/whats_new/v1.0.rst:
```diff
@@ -217,6 +217,19 @@ Changelog
   :pr:`19426` by :user:`Alexandre Gramfort <agramfort>` and
   :user:`Maria Telenczuk <maikia>`.
 
+- |Fix| :class:`RANSACRegressor` Points with a residual equal to 
+  ``residual_threshold`` member are considered as an inlier.
+  Motivation for this change: We want a perfect dataset to be considered as 
+  inliers, specifically with default parameters or ``residual_threshold`` 
+  equal to zero.
+  Previously the code launched an exceptions with perfect model.
+  Such a comportement also happen with MAD and half of the 
+  points perfect.
+  With this definition, ``residual_threshold`` set to zero no longer set all 
+  points to outliers. Now you can do that by setting ``residual_threshold`` to 
+  nan.
+  :pr:`19499` by :user:`Gregory Strubel <gregorystrubel>`.
```
I think this can be made shorter: readers interested by the details of those changes can refer to this PR.

```suggestion
- |Fix| Points with a residual equal to  ``residual_threshold`` are now considered
  as an inliers for :class:`linear_model.RANSACRegressor`. This allows fitting
  model perfectly on some datasets when ``residual_threshold=0``.
  :pr:`19499` by :user:`Gregory Strubel <gregorystrubel>`.
```

jjerphan commented on sklearn/linear_model/_ransac.py:
```diff
@@ -95,7 +95,8 @@ class RANSACRegressor(MetaEstimatorMixin, RegressorMixin,
         Maximum residual for a data sample to be classified as an inlier.
         By default the threshold is chosen as the MAD (median absolute
-        deviation) of the target values `y`.
+        deviation) of the target values `y`. Residual strictly equal to the
+        threshold is considered valid.
```
```suggestion
        deviation) of the target values `y`. Points whose residuals are
        strictly equal to the threshold are considered as inliers.
```

jjerphan reviewed (changes requested):
Thanks for this PR @gregorystrubel! Here are a few suggestions.

gregorystrubel committed:
Update doc/modules/linear_model.rst

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Commit:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into bugfix_ransac_residual_threshold

Commit:
precise "non-regression" test

Relative to jjerphan comment
https://github.com/scikit-learn/scikit-learn/pull/19499#discussion_r648013787

gregorystrubel committed:
Update doc/whats_new/v1.0.rst

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Commit:
Merge branch 'bugfix_ransac_residual_threshold' of github.com:gregorystrubel/scikit-learn into bugfix_ransac_residual_threshold

gregorystrubel committed:
Update sklearn/linear_model/_ransac.py

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

gregorystrubel commented:
Thanks @jjerphan, additionally to your suggestions, I precised "non-regression" in the comment of the non-regression test 
(commit #ddacfe2)

jjerphan reviewed (approved):
LGTM, thank you @gregorystrubel!

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/gregorystrubel/19499

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -456,6 +456,11 @@ Changelog
   :user:`Maria Telenczuk <maikia>`.
 
+- |Fix| Points with a residual equal to  ``residual_threshold`` are now considered
+  as an inliers for :class:`linear_model.RANSACRegressor`. This allows fitting
```
```suggestion
  as inliers for :class:`linear_model.RANSACRegressor`. This allows fitting
```

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -456,6 +456,11 @@ Changelog
   :pr:`19426` by :user:`Alexandre Gramfort <agramfort>` and
   :user:`Maria Telenczuk <maikia>`.
 
+- |Fix| Points with a residual equal to  ``residual_threshold`` are now considered
```
```suggestion
- |Fix| Points with residuals equal to  ``residual_threshold`` are now considered
```

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -456,6 +456,11 @@ Changelog
 
+- |Fix| Points with a residual equal to  ``residual_threshold`` are now considered
+  as an inliers for :class:`linear_model.RANSACRegressor`. This allows fitting
+  model perfectly on some datasets when ``residual_threshold=0``.
```
```suggestion
  a model perfectly on some datasets when `residual_threshold=0`.
```

glemaitre commented on sklearn/linear_model/tests/test_ransac.py:
```diff
@@ -169,13 +169,13 @@ def test_ransac_predict():
 
 
 def test_ransac_resid_thresh_no_inliers():
```
```suggestion
def test_ransac_residuals_threshold_no_inliers():
```

glemaitre commented on sklearn/linear_model/tests/test_ransac.py:
```diff
@@ -597,6 +597,19 @@ def test_ransac_final_model_fit_sample_weight():
 
 
+def test_perfect_horizontal_line():
+    # non-regression test for issue #19497
```
```suggestion
    """Check that we can fit a line where all samples are inliers.
    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/issues/19497
    """
```

glemaitre committed:
Apply suggestions from code review

glemaitre commented:
I merged my small suggestions that were only typo. We will check if the tests are passing and we will be able to merge.
